### PR TITLE
Add ResultIterable<T>.map(Function<T, U>) method

### LINKED ISSUE #1527: ResultIterable<T> to ResultIterable<V>
First off, I love jdbi3. I have been using it a while now because it is such a pleasure to use, so thank you very much for the library!

I am wondering if it has been considered to provide a method which maps a `ResultIterable<T>` to `ResultIterable<V>` perhaps given a `Function<T, V>`.

My use case is that I have a "low level" database interface that ideally would return a `ResultIterable<T>`.  Callers of that interface may want to transform that `T` into a different "higher level" type `V` without having to deal with `Stream<T>` (somewhat verbose via `withStream`, and alternatives have their flaws too).

I could go into the details about why it seems helpful, or maybe what the actual implementation would look like, but I first wanted to know if this had been considered or is infeasible for some reason.

### PR DESCRIPTION
**Pull request by qualidafial**
Fixes #1527

### CODE REVIEW

**Comment by qualidafial** [comment]
Build failing from bogus PMD error--wants me to name an instance field like a constant.

**Comment by qualidafial** [comment]
Opened an issue with PMD: pmd/pmd#1812

File: core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java [hunk 1]
```diff
@@ -81,6 +82,40 @@
...
 
+    /**
+     * Returns a {@code ResultIterable<U>} derived from this {@code ResultIterable<T>}, by
+     * transforming elements using the given mapper function.
+     *
+     * @param mapper function to apply to elements of this ResultIterable
+     * @param <U>    Element type of the returned ResultIterable
+     * @return the new ResultIterable
+     */
+    default <U> ResultIterable<U> map(Function<T, U> mapper) {
```
Comment (thekeenant) [inline core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java]: Would `Function<? super T, ? extends U>` be more appropriate? This is convenient for passing `Function`'s, or specifically method references, which are compatible with the `T` input and `U` output but are not strictly those types, no casting needed ([Stream#map](https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/java/util/stream/Stream.java#L184)).

Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java]: Done

**Comment by qualidafial** [comment]
Closed PMD issue--upgrading to latest PMD version resolved it.

**Comment by qualidafial** [comment]
I think I may have inadvertently introduced an infinite recursion. Will resolve once I'm off the road

**Comment by qualidafial** [comment]
@stevenschlansker Yep, the only thing we needed to change was where (and when) the delegate Iterable gets created.

File: core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java [hunk 2]
```diff
@@ -81,6 +82,40 @@
...
+     * Returns a {@code ResultIterable<U>} derived from this {@code ResultIterable<T>}, by
+     * transforming elements using the given mapper function.
+     *
+     * @param mapper function to apply to elements of this ResultIterable
+     * @param <U>    Element type of the returned ResultIterable
+     * @return the new ResultIterable
+     */
+    default <U> ResultIterable<U> map(Function<? super T, ? extends U> mapper) {
+        return () -> new ResultIterator<U>() {
+            ResultIterator<T> delegate = iterator();
```
Comment (TheRealMarnes) [inline core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java]: should be `private final`, no? The return signature will prevent it from being accessed, but still, if it was a class in its own right we wouldn't allow it.

Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java]: Yep. I'll push another change

Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java]: Addressed in #1530

**Pull request merged**